File: 백준/Gold/2234. 성곽/성곽.cc
#include <bits/stdc++.h>

using namespace std;

int m, n, cnt, a[54][54], visited[54][54], compSize[2504], mx, big;
const int dy[] = {0, -1, 0, 1};
const int dx[] = {-1, 0, 1, 0};
int go(int y, int x, int cnt) {
    if(visited[y][x]) return 0;
    visited[y][x] = cnt;
    int ret = 1;
    for(int i = 0; i < 4; i++) {
        if(!((1 << i) & a[y][x])) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if(ny < 0 || ny >= m || nx < 0 || nx >= n) continue;
            ret += go(ny, nx, cnt);
        }
    }
    return ret;
}

int main() {
    cin >> n >> m;
    for(int i = 0; i < m; i++) {
        for(int j = 0; j < n; j++) {
            cin >> a[i][j];
        }
    }
    for(int i = 0; i < m; i++) {
        for(int j = 0; j < n; j++) {
            if(!visited[i][j]) {
                cnt++;
                compSize[cnt] = go(i, j, cnt);
                mx = max(mx, compSize[cnt]);
            }
        }
    }
    
    for(int i = 0; i < m; i++) {
        for(int j = 0; j < n; j++) {
            if(i + 1 < m) {
                int a = visited[i][j];
                int b = visited[i + 1][j];
                if(a != b) big = max(big, compSize[a] + compSize[b]);
            }
            if(j + 1 < n) {
                int a = visited[i][j];
                int b = visited[i][j + 1];
                if(a != b) big = max(big, compSize[a] + compSize[b]);
            }
        }
    }
    
    cout << cnt << '\n' << mx << '\n' << big << '\n';
    return 0;
}

File: 백준/Gold/2234. 성곽/성곽.swift
import Foundation

func solve() {
    let dy = [0, -1, 0, 1]
    let dx = [-1, 0, 1, 0]
    let input = readLine()!.split(separator: " ").map { Int($0)! }
    let (m, n) = (input[0], input[1])
    var walls = [[Int]]()
    var visited = Array(repeating: Array(repeating: 0, count: m), count: n)
    var sizes = [Int]()
    for _ in 0..<n {
        let wall = readLine()!.split(separator: " ").map { Int($0)! }
        walls.append(wall)
    }

    var cnt = 0
    var maxSize = 0
    var big = 0

    func dfs(_ y: Int, _ x: Int, number: Int) -> Int {
        if visited[y][x] != 0 { return 0 }
        visited[y][x] = number
        var size: Int = 1

        for i in 0..<4 {
            if walls[y][x] & (1 << i) != 0 { continue }
            let ny = y + dy[i]
            let nx = x + dx[i]
            size += dfs(ny, nx, number: number)
        }

        return size
    }

    for i in 0..<n {
        for j in 0..<m {
            if visited[i][j] == 0 {
                cnt += 1
                let curSize = dfs(i, j, number: cnt)
                sizes.append(curSize)
                maxSize = max(maxSize, curSize)
            }
        }
    }
    
    for i in 0..<n {
        for j in 0..<m {
            if (i + 1 < n) {
                let a = visited[i][j]
                let b = visited[i + 1][j]

                if a != b {
                    big = max(big, sizes[a - 1] + sizes[b - 1])
                }
            }
            if (j + 1 < m) {
                let a = visited[i][j]
                let b = visited[i][j + 1]

                if a != b {
                    big = max(big, sizes[a - 1] + sizes[b - 1])
                }
            }
        }
    }

    print(cnt)
    print(maxSize)
    print(big)
}

solve()